### PR TITLE
Add latest cqc rating

### PR DESCRIPTION
**Pull request by RoyPriceCS50**
# Description
Added two functions to merge_coverage_data job. One for filtering cqc ratings to latest only. One for applying the filter and then joining the ratings to coverage dataframe.
The join adds the latest overall rating and the rating date to the coverage dataframe.

Also changed the cqc ratings column name script so that date = "rating_date" instead of "date".

Link to Trello:  https://trello.com/c/uWWQi5a3

# How to test
Unit tests are passing.

Link to successful run in Glue: https://eu-west-2.console.aws.amazon.com/gluestudio/home?region=eu-west-2#/editor/job/add-latest-cqc-rating-merge_coverage_data_job/runs

Link to successful ind cqc filled posts step function: https://eu-west-2.console.aws.amazon.com/states/home?region=eu-west-2#/v2/executions/details/arn:aws:states:eu-west-2:344210435447:execution:add-latest-cqc-rating-Ind-CQC-Filled-Post-Estimates-Pipeline:5994db55-d866-4fbb-90c5-6f72e384a198

# Developer checklist
- [x] Unit test
- [x] Linked to Trello ticket
- [ ] Documentation up to date

### CODE REVIEW

File: utils/column_values/categorical_column_values.py [hunk 1]
```diff
@@ -461,6 +461,8 @@ class CQCRatingsValues(ColumnValues):
     historic: str = "Historic"
     outstanding: str = "Outstanding"
     good: str = "Good"
+    latest_rating: int = 1
```
Comment (MarionJHolloway) [inline utils/column_values/categorical_column_values.py:464]: So these are from a different column in the same dataset. They will need a new class for that column.

File: jobs/merge_coverage_data.py [hunk 2]
```diff
@@ -173,13 +191,68 @@ def add_flag_for_in_ascwds(
     return merged_coverage_df
 
 
+def keep_only_latest_cqc_ratings(
```
Comment (MarionJHolloway) [inline jobs/merge_coverage_data.py:194]: It's generally good practice to use descriptive verbs for function names. Most of our functions designed to filter start with "filter_..." It just makes it easier to scan the code and understand what transformation has been performed.

File: jobs/merge_coverage_data.py [hunk 3]
```diff
@@ -173,13 +191,68 @@ def add_flag_for_in_ascwds(
...
+        == CQCRatingsValues.latest_rating
+    )
+
+
+def join_latest_cqc_rating_into_coverage_df(
+    merged_coverage_df: DataFrame,
+    cqc_ratings_df: DataFrame,
+) -> DataFrame:
+    """
+     Join latest rating to coverage dataframe using locationid as key.
```
Comment (MarionJHolloway) [inline jobs/merge_coverage_data.py:222]: Formatting - you have a space at the start of the comment

File: terraform/pipeline/glue.tf [hunk 4]
```diff
@@ -393,6 +393,7 @@ module "merge_coverage_data_job" {
   job_parameters = {
     "--cleaned_cqc_location_source"         = "${module.datasets_bucket.bucket_uri}/domain=CQC/dataset=locations_api_cleaned/"
     "--workplace_for_reconciliation_source" = "${module.datasets_bucket.bucket_uri}/domain=SfC/dataset=workplace_for_reconciliation/"
+    "--cqc_ratings_source"                  = "${module.datasets_bucket.bucket_uri}/domain=SfC/dataset=cqc_ratings/"
```
Comment (MarionJHolloway) [inline terraform/pipeline/glue.tf:396]: You will need to update the step function in terraform too!

File: tests/unit/test_merge_coverage_data.py [hunk 5]
```diff
@@ -122,5 +133,70 @@ def test_add_flag_for_in_ascwds_has_expected_values(self):
         self.assertEqual(returned_rows, expected_rows)
 
 
+class KeepOnlyLatestCqcRatingTests(unittest.TestCase):
+    def setUp(self) -> None:
+        self.spark = utils.get_spark()
+
+        self.sample_cqc_ratings_df = self.spark.createDataFrame(
```
Comment (MarionJHolloway) [inline tests/unit/test_merge_coverage_data.py:140]: Is this not the same test data and schema as in row 44? Best to create it once and reuse.

File: tests/unit/test_merge_coverage_data.py [hunk 6]
```diff
@@ -122,5 +133,70 @@ def test_add_flag_for_in_ascwds_has_expected_values(self):
         self.assertEqual(returned_rows, expected_rows)
 
 
+class KeepOnlyLatestCqcRatingTests(unittest.TestCase):
```
Comment (MarionJHolloway) [inline tests/unit/test_merge_coverage_data.py:136]: Change this to inherit from MergeCoverageDatasetTests (and the other test class you made) so you don't have to duplicate your data :)

File: tests/test_file_data.py [hunk 7]
```diff
@@ -2247,6 +2247,22 @@ class MergeCoverageData:
         ("1", 1),
     ]
 
+    sample_cqc_locations_rows = [("1-000000001",), ("1-000000002",)]
+
+    sample_cqc_ratings_for_merge_rows = [
+        ("1-000000001", "2024-01-01", "Good", 0),
+        ("1-000000001", "2024-01-02", "Good", 1),
+        ("1-000000001", None, "Good", None),
```
Comment (MarionJHolloway) [inline tests/test_file_data.py:2255]: Well done for testing with null values

File: tests/unit/test_merge_coverage_data.py [hunk 8]
```diff
@@ -122,5 +133,70 @@ def test_add_flag_for_in_ascwds_has_expected_values(self):
...
+        self.returned_in_ascwds_df = job.keep_only_latest_cqc_ratings(
+            self.sample_cqc_ratings_df
+        )
+
+    def test_keep_only_latest_cqc_ratings_only_contains_latest_ratings(self):
+        latest_rating_column_df = self.returned_in_ascwds_df.select(
+            CQCRatingsColumns.latest_rating_flag
+        )
+
+        distinct_latest_rating_rows = latest_rating_column_df.distinct().collect()
```
Comment (MarionJHolloway) [inline tests/unit/test_merge_coverage_data.py:154]: Why do you use distinct here?

File: tests/unit/test_merge_coverage_data.py [hunk 9]
```diff
@@ -122,5 +133,70 @@ def test_add_flag_for_in_ascwds_has_expected_values(self):
...
+    def test_join_latest_cqc_rating_into_coverage_df_adds_expected_columns(self):
+        self.assertEqual(len(self.returned_df.columns), len(self.expected_df.columns))
+
+    def test_join_latest_cqc_rating_into_coverage_df_does_not_add_any_rows(self):
+        self.assertEqual(self.returned_df.count(), self.expected_df.count())
+
+    def test_join_latest_cqc_rating_into_coverage_df_has_no_duplicate_columns(self):
+        self.assertEqual(
+            sorted(self.returned_df.columns),
+            sorted(list(set(self.returned_df.columns))),
```
Comment (MarionJHolloway) [inline tests/unit/test_merge_coverage_data.py:194]: Nice!

**Review by MarionJHolloway — CHANGES_REQUESTED** [review]
Well done! You're skills are really improving :)

Happy to chat through the comments on a call if that's easier.

File: tests/unit/test_merge_coverage_data.py (hunk 8 above)
Comment (RoyPriceCS50) [reply tests/unit/test_merge_coverage_data.py:154]: The dataframe has only 1 column, the latest rating flag.
Distinct removes duplicate values, so I want to only see 1's remaining. If that's true, then I only have 1 row left and the value of that row is 1.

File: tests/unit/test_merge_coverage_data.py [hunk 10]
```diff
@@ -122,5 +143,63 @@ def test_add_flag_for_in_ascwds_has_expected_values(self):
...
+        )
+
+        distinct_latest_rating_rows = latest_rating_column_df.distinct().collect()
+
+        self.assertEqual(len(distinct_latest_rating_rows), 1)
+        self.assertEqual(
+            distinct_latest_rating_rows[0][0], CQCLatestRating.is_latest_rating
+        )
+
+    def test_filter_for_latest_cqc_ratings_only_has_expected_row_count(self):
```
Comment (MarionJHolloway) [inline tests/unit/test_merge_coverage_data.py:166]: I don't think you need "only" in this test name :)

**Review by MarionJHolloway — CHANGES_REQUESTED** [review]
Because you've updated the step function, can you run the step function in branch to test that's all wired up properly and add a link to it in the PR description. There's also one small tweak to wording in the tests.  After that, I can approve :)

**Review by MarionJHolloway — APPROVED** [review]
Well done!

**Pull request merged**